src/netbase.cpp

```diff
@@ -554,8 +567,6 @@ bool ConnectThroughProxy(const proxyType &proxy, const std::string& strDest, int
         if (!Socks5(strDest, (unsigned short)port, 0, hSocket))
             return false;
     }
-
-    hSocketRet = hSocket;
```

Comment:
If socks negotiation fail then `hSocket` will not be `INVALID_SOCKET` which is a different behavior than before. This also indicates that there is a leak in master because `hSocket` is not freed?

Reply:
Nice Catch! Force-pushed the missing CloseSocket()s.
```diff
diff --git a/src/netbase.cpp b/src/netbase.cpp
index 18763d2..4b72826 100644
--- a/src/netbase.cpp
+++ b/src/netbase.cpp
@@ -561,11 +561,15 @@ bool ConnectThroughProxy(const proxyType &proxy, const std::string& strDest, int
         ProxyCredentials random_auth;
         static std::atomic_int counter(0);
         random_auth.username = random_auth.password = strprintf("%i", counter++);
-        if (!Socks5(strDest, (unsigned short)port, &random_auth, hSocket))
+        if (!Socks5(strDest, (unsigned short)port, &random_auth, hSocket)) {
+            CloseSocket(hSocket);
             return false;
+        }
     } else {
-        if (!Socks5(strDest, (unsigned short)port, 0, hSocket))
+        if (!Socks5(strDest, (unsigned short)port, 0, hSocket)) {
+            CloseSocket(hSocket);
             return false;
+        }
     }
     return true;
 }
```

Reply:
`CloseSocket` sets `hSocket` to `INVALID_SOCKET` so this LGTM. I wonder if you could just `return CloseSocket(hSocket)`?

Reply:
CloseSocket's return value indicates whether the close succeed or not, which would be true in this case. Not what we want :)

Reply:
Ahhh yeah... now I win the most stupid comment award.

Reply:
Hah, no worries. We'll call it even for pointing out my mistake here